libs/wingc/src/type_check.rs

```diff
@@ -5793,9 +5900,97 @@ impl<'a> TypeChecker<'a> {
 						} else {
 							v.phase
 						};
-						(v.clone(), phase)
+						(ResolveReferenceResult::Variable(v.clone()), phase)
+					}
+					_ => {
+						let err = self.spanned_error_with_var(property, format!("\"{}\" not a valid reference", reference));
+						(ResolveReferenceResult::Variable(err.0), err.1)
+					}
+				}
+			}
+			Reference::ElementAccess { object, index } => {
+				let (instance_type, instance_phase) = self.type_check_exp(object, env);
+				let (index_type, _index_phase) = self.type_check_exp(index, env);
+
+				match *instance_type {
+					// TODO: it might be possible to look at Type::Json's inner data to give a more specific type
+					Type::Json(_) => {
+						self.validate_type_in(index_type, &[self.types.number(), self.types.string()], index);
+						(
+							ResolveReferenceResult::Location(instance_type, self.types.json()), // indexing into a Json object returns a Json object
+							instance_phase,
+						)
+					}
+					Type::MutJson => {
+						self.validate_type_in(index_type, &[self.types.number(), self.types.string()], index);
+						(
+							ResolveReferenceResult::Location(instance_type, self.types.mut_json()), // indexing into a MutJson object returns a MutJson object
+							instance_phase,
+						)
+					}
+					Type::Array(inner_type) => {
+						self.validate_type(index_type, self.types.number(), index);
+						(
+							ResolveReferenceResult::Location(instance_type, inner_type),
+							instance_phase,
+						)
+					}
+					Type::MutArray(inner_type) => {
+						self.validate_type(index_type, self.types.number(), index);
+						(
+							ResolveReferenceResult::Location(instance_type, inner_type),
+							instance_phase,
+						)
+					}
+					Type::Map(inner_type) => {
+						self.validate_type(index_type, self.types.string(), index);
+						(
+							ResolveReferenceResult::Location(instance_type, inner_type),
+							instance_phase,
+						)
+					}
+					Type::MutMap(inner_type) => {
+						self.validate_type(index_type, self.types.string(), index);
+						(
+							ResolveReferenceResult::Location(instance_type, inner_type),
+							instance_phase,
+						)
+					}
+					Type::Anything => (
+						ResolveReferenceResult::Location(self.types.anything(), self.types.anything()),
+						instance_phase,
+					),
+					Type::Unresolved => (
+						ResolveReferenceResult::Location(self.types.error(), self.types.error()),
+						instance_phase,
+					),
+					Type::Inferred(_) => (
+						ResolveReferenceResult::Location(self.types.error(), self.types.error()),
```

Comment:
Is this really an error? Not sure how this works exactly.

Reply:
I wasn't sure how to get inferred types working here so for now I've made it an error (I've added a test case just now). Open to improving this